cranelift-codegen/meta/src/isa/x86/legalize.rs

```diff
@@ -290,4 +297,80 @@ pub fn define(shared: &mut SharedDefinitions, x86_instructions: &InstructionGrou
     );
 
     group.build_and_add_to(&mut shared.transform_groups);
+
+    let mut narrow = TransformGroupBuilder::new(
+        "x86_narrow",
+        r#"
+    Legalize instructions by narrowing.
+
+    Use x86-specific instructions if needed."#,
+    )
+    .isa("x86")
+    .chain_with(shared.transform_groups.by_name("narrow").id);
+
+    // SIMD
+    let broadcast_immediate_u8 = Literal::constant(uimm8, 0x00);
+    // let broadcast_immediate_f64 = Literal::constant(ieee64, 0x00);
+    // let lane_zero = Literal::constant(uimm8, 0x00);
+    let lane_one = Literal::constant(uimm8, 0x01);
+
+    // SIMD splat: 8-bits
+    //    narrow.legalize(
+    //        def!(y = splat.I8x16(x)), // TODO this is not possible yet
+    //        vec![
+    //            def!(a = bitcast(x)), // move into the lowest 8 bits of an XMM register TODO this encoding does not yet exist
+    //            def!(b = f64const(broadcast_immediate_f64)), // setup the shuffle mask in another XMM register; this only works if f64const zeroes out the top 64 bits TODO verify this mask is correct
+    //            def!(y = x86_pshufb(a, b)), // PSHUFB takes two XMM operands, one of which is a shuffle mask (i.e. b); TODO this encoding does not exist yet
+    //        ],
+    //    );
+
+    // SIMD splat: 16-bits
+    //    let b = var("b");
+    //    let c = var("c");
+    //    let d = var("d");
+    //    for ty in LaneTypeIterator::new().filter(|t| t.lane_bits() == 16) {
+    //        let expr = ExprBuilder::apply(splat.bind_vector(ty, 8).into(), vec![x.clone().into()])
+    //            .assign_to(vec![y.clone().into()]);
+    //        let expr_bitcast_i32x4 =
+    //            ExprBuilder::apply(bitcast.bind_vector(I32, 4).into(), vec![b.clone().into()])
+    //                .assign_to(vec![c.clone().into()]);
+    //        let expr_bitcast_x16x8 =
+    //            ExprBuilder::apply(bitcast.bind_vector(ty, 8).into(), vec![d.clone().into()])
+    //                .assign_to(vec![y.clone().into()]);
+    //
+    //        narrow.legalize(
+    //            expr, // def!(y = splat.X16x8(x)),
+    //            vec![
+    //                def!(a = x86_scalar_to_vector(x)), // translate to an x86 MOV to get the value in the lowest 16 bits of an XMM register
+    //                def!(b = insertlane(a, lane_one, x)), // insert the value again but in the next lowest 16 bits
+    //                expr_bitcast_i32x4, // def!(c = bitcast.I32#4(b)), // pretend this is an I32x4 so we can use PSHUFD, this should translate to a NOOP
+    //                def!(d = x86_pshuf(c, broadcast_immediate_u8)), // broadcast the bytes in the XMM register with PSHUF
+    //                expr_bitcast_x16x8, // def!(y = bitcast.ty#8(d)), // pretend this is an X16x8 again, this should translate to a NOOP
```

Comment:
Since there is no good instruction for shuffling 16-bit lanes in x86 (`PSHUFW` seems to only work for MM registers) my plan was to fill up the bottom 32 bits, "cast" to a 32x4, shuffle with the 32x4 PSHUFD, and then "cast" back. I can't get the above to work in cranelift due to some typing issues. Is this type of "cast" possible? What is the best way to tell cranelift to cast for SSA purposes but emit no instructions?

Reply:
Another option for 16-bit lane shuffling is to use `pshufb`, when `SSSE3` is available.

`bitcast` ought to be the way to do this, and emit zero instructions when possible. If there are problems using `bitcast` for this, they're bugs and we should fix them :-).

Reply:
For what it's worth, for 16-bit lane splat, in Spidermonkey, we used to duplicate the content of the low 8 bits to the low 16 bits (with a shift and OR), and then do a 8-bit splat, if SSSE3 is not available. This sounds like what you want to achieve here with a 32-bit splat.

Re: `bitcast`, approving what Dan says. In addition to this, you might need to add a few encodings for some bound types, with a recipe that emits no code (like the `null` recipe for instance, but might need a new one for FPR / vector types).

Reply:
Thanks for the help on this; I implemented it with the 32-bit PSHUFD (supported in SSE2) because I think it can be done with 3 instructions (MOV + PINSRD + PSHUFD) but I can switch over to PSHUFB if that doesn't work. Let me comment over on the new version of this; there's a test failing...